File: src/main/scala/98.validate-binary-search-tree.scala
/**
  * Definition for a binary tree node.
  * class TreeNode(var _value: Int) {
  * var value: Int = _value
  * var left: TreeNode = null
  * var right: TreeNode = null
  * }
  */
object Q98 {
  import common.scala.TreeNode
  def isValidBST(root: TreeNode): Boolean = {
    def loop(root: TreeNode, low: Int, high: Int): Boolean =
      if (root == null) true
      else if (!(root.value > low && root.value < high)) false
      else loop(root.left, low, root.value) && loop(root.right, root.value, high)
    if (root == null) true
    else loop(root, root.value, root.value)
  }
}

File: src/main/java/98.validate-binary-search-tree.java
import java.util.Stack;

import common.java.TreeNode;

/**
 * Definition for a binary tree node. public class TreeNode { int val; TreeNode
 * left; TreeNode right; TreeNode(int x) { val = x; } }
 *
 * <pre>
 * 思路:
 * 依赖迭代中序遍历, 并且中序遍历的结果为从小到大排列
 * 比较当前节点跟上一个节点的大小
 * </pre>
 */
class Q98 {
  public boolean isValidBST(TreeNode root) {
    Stack<TreeNode> stack = new Stack<>();
    TreeNode curr = root;
    TreeNode prev = null;
    while (!stack.isEmpty() || curr != null) {
      while (curr != null) {
        stack.push(curr);
        curr = curr.left;
      }
      TreeNode top = stack.pop();
      if (prev != null && top.val <= prev.val) {
        return false;
      }
      prev = top;
      curr = top.right;
    }
    return true;
  }
}
